### PR TITLE
Split jar execution strategy for client and member [HZ-2097 HZ-2058]

### PR DESCRIPTION
**Pull request by orcunc**
The implementation options are
1. Use thread local object to store the parameters. When the main method in the jar executes,  since it will still be in the same thread, the parameters can be accessed. This implementation does not require locking but requires that the main method submits the job from the invoking thread
2. Reduce the scope of the synchronized block
3. Pass a unique key as main method argument. Then the user can call newJob(... JobConfig) overloads as a callback . The key can be extracted from the JobConfig and parameters can be used

Option 1 puts constraint on the client. 
Option 2 has the risk of blocking the member if job gets stuck in the main method.  Option 1 also has this problem but it will only block one thread. I am not sure what to do with the blocked thread.  
Option 3 is very limiting. 

Also the jars executed by the client and member need to use different strategies. 
For the client
- The execute operation should not be a synchronized method. Because there is only one client and the HZ instance is shutdown after the jar is executed

For the member currently the PR is doing option 2. 
- The execute operation should use a smaller synchronized block. Because the member does not need to wait for the job to start

Fixes : 
https://github.com/hazelcast/hazelcast/issues/23781 - confusing exception
https://github.com/hazelcast/hazelcast/issues/23664 - synchronized method 

Jira : 
https://hazelcast.atlassian.net/browse/HZ-2097 - confusing exception
https://hazelcast.atlassian.net/browse/HZ-2058 - synchronized method

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

**Comment by orcunc** [comment]
run-sonar

**Review by olukas — CHANGES_REQUESTED** [review]
This PR doesn't fix https://github.com/hazelcast/hazelcast/issues/23664 at all. Main method is still executed in synchronized block. It still means that if something gets stuck in main method of submitted Jar then the whole jar upload is blocked forever (and cannot heal up until HZ node is restarted).

We have to be able to run more main methods from different jars at the same time.

(I haven't reviewed the rest of PR since this is something what needs to be reimplemented)

File: hazelcast/src/main/java/com/hazelcast/instance/impl/executejar/ClientExecuteJarStrategy.java
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/instance/impl/executejar/ClientExecuteJarStrategy.java:52]: I don't understand why this mean that the method must be synchronized.

You can do `BootstrappedInstanceProxy instanceProxy = singleton.remembered();` before try block, then reference will be constant for this method - no need for any synchrionization

**Comment by orcunc** [comment]
> This PR doesn't fix #23664 at all. Main method is still executed in synchronized block. It still means that if something gets stuck in main method of submitted Jar then the whole jar upload is blocked forever (and cannot heal up until HZ node is restarted).
> 
> We have to be able to run more main methods from different jars at the same time.
> 
> (I haven't reviewed the rest of PR since this is something what needs to be reimplemented)

This is the link to the discussion in the slack channel
https://hazelcast.slack.com/archives/C04ERT4SLBC/p1677844095922569

So far no other suggestions. As I explained in the options, at least for now let's reduce the scope of the synchronized block. Yes @olukas you are right if something gets stuck, the thread will not be used anymore. But 
1. Only one thread will be lost. The member will still function with other threads. 
2. This is an internal API and the job submitting person should be careful not to submit dangerous jobs
3. I prefer to pass all parameters to the jar and not to store any state in the HazelcastBootsrap class. We can define a parameter convention for this. But all parties must agree on this. And so far no response.

**Comment by olukas** [comment]
@orcunc 

> Only one thread will be lost. The member will still function with other threads.

if main method is run in synchronized block and it get stucks there that it's not true that it will be able to run in other threads. It will block forever due to synchronized block.

> This is an internal API and the job submitting person should be careful not to submit dangerous jobs

How do you define "dangerous jobs"? I guess that no user will think about it until it will get stuck on some day eventually (for example because it will read something from database which will be unreachable without configured timeout).

It's not related to public/private API - if we will have it in CLC and user can just send a jar by this feature then it's totally irrelevant it's private API.

> I prefer to pass all parameters to the jar and not to store any state in the HazelcastBootsrap class. We can define a parameter convention for this. But all parties must agree on this. And so far no response.

Implementation should be engineering decision and not QE decision. As a QE I can tell you what's wrong (i.e. if main method gets stuck it will make the feature not working forever) but I cannot tell you how to fix it.

**Review by olukas — APPROVED** [review]
**PLEASE DON'T TAKE THIS APPROVE AS REAL APPROVE.**

Setting as "approve" because I requested changes (but they will not be probably addressed by this PR) and it will not allow to merge until I set it as "approve" even if I would like remove myself from the reviewer list.

If https://github.com/hazelcast/hazelcast/issues/23664 would be addressed by this PR then please set me again as reviewer and I check whether it fixes the issue.

**Pull request closed without merging**
